prov/gni/src/gnix_buddy_allocator.c

```diff
@@ -0,0 +1,353 @@
+		│ BASE │ len  │ min │ max │ nlists │ LISTS │
+		└──────┴──────┴─────┴─────┴────────┴───────┘
+ * The LISTS pointer points to an array of dlist structures, each containing a
+ * head pointer to the begging of a free list.  Note that the first element of
+ * LISTS is a pointer to the head of the "min block size" free list, the second
+ * element is the head of the "min * 2 block size" free list and so on.
+
+ 		Node format as stored in a free block:
+		┌──────┬──────┬──────────────────────┐
+		│ NEXT │ PREV │ Remaining free bytes │
+		└──────┴──────┴──────────────────────┘
+ * Each NEXT and PREV pointer is stored in the first 16 bytes of the free block.
+ * This means that there is a hard limit of 16 bytes on the minimum block size.
+ *
+ *		Bitmap layout with a min block size of 16:
+ 		┌──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┐
+		│16│16│16│16│..│32│32│32│32│..│64│64│64│64│..│
+		└──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┘
+ * Each number above (16, 32, and 64) represents a block size that can
+ * be split. The first two "32" blocks are redundantly mapped by the the first
+ * 64 block, all blocks are redundant mappings of the next block size.
+ *
+ * When a block is split or allocated, its bit will be set in the bitmap.
+ * When a block is coalesced or free'd its bit will be reset in the bitmap.
+ * This improves the performance of coalescing blocks by being able to tell
+ * whether a given block's buddy block is allocated or split in O(c).
+ *
+ * TODO: dlist_insert_sorted for fragmentation reduction.
+ * TODO: Lock in __gnix_buddy_split and allow __gnix_buddy_find_block to run
+ * concurrently.
+ */
+
+#include "gnix_buddy_allocator.h"
+
+static inline int __gnix_buddy_create_lists(handle_t *alloc_handle)
+{
+	size_t i, offset = 0;
+
+	alloc_handle->nlists = (size_t) log2(alloc_handle->max /
+					     (double) MIN_BLOCK_SIZE) + 1;
+	alloc_handle->lists = calloc(1, sizeof(struct dlist_entry) *
+				     alloc_handle->nlists);
+
+	if (!alloc_handle->lists) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not create buddy allocator lists.\n");
+		return -FI_ENOMEM;
+	}
+
+	for (i = 0; i < alloc_handle->nlists; i++) {
+		dlist_init(alloc_handle->lists + i);
+	}
+
+	/* Insert free blocks of size max in sorted order into last list */
+	for (i = 0; i < alloc_handle->len / alloc_handle->max; i++) {
+		dlist_insert_tail(alloc_handle->base + offset,
+				  alloc_handle->lists +
+				  alloc_handle->nlists - 1);
+		offset += alloc_handle->max;
+	}
+
+	return FI_SUCCESS;
+}
+
+/**
+ * Split a block in list "j" until list "i" is reached.
+ */
+static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
+{
+	void *tmp = NULL;
+
+	dlist_remove(tmp = alloc_handle->lists[j].next);
+
+	_gnix_set_bit(&alloc_handle->bitmap,
+		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
+				   OFFSET(MIN_BLOCK_SIZE, j)));
+
+
+	/* Split the block until we reach list "i" */
+	for (j--; j > i; j--) {
```

Comment:
Sure, the bitmap in the handle_t struct has a bit for every block size, the smallest blocks are mapped to in the first half of the bitmap and the rest of the blocks are mapped to in the second half of the bitmap.  The bitmap is over allocated and will always waste at least one bit.

A bit (corresponding to a block of a certain size) is set when that block is allocated or split.

For example, if the length of the base block is 128B,
MinBlockSize = 16B,
MaxBlockSize = 64B then bitmap looks like:

16|16|16|16|16|16|16|16|32|32|32|32|64|64

So, all the 16s are bits mapping to all the blocks of size 16 that we can fit in our base block; this goes for the 32s and 64s as well.  We are not using the 2 hidden bits after the last 64.

When freeing, we try to coalesce the free block with a buddy block; the two blocks must be the same size.  We can find the state of any buddy block quickly with BITMAP_INDEX and if that bit is set, then we have to insert the free block into its free list without coalescing.  If the bit is not set, we coalesce until we reach a buddy block that is allocated or split, or we reach the MaxBlockSize list, then we insert the block into its free list.

So, by using this bitmap layout we avoid searching through our free lists but waste some space.

> On Jan 7, 2016, at 10:46 PM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_buddy_allocator.c:
> 
> > +/**
> > + * Split a block in list "j" until list "i" is reached.
> > + */
> > +static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
> > +{
> > +	void *tmp = NULL;
> > +
> > +	dlist_remove(tmp = alloc_handle->lists[j].next);
> > +
> > +	_gnix_set_bit(&alloc_handle->bitmap,
> > +		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
> > +				   OFFSET(MIN_BLOCK_SIZE, j)));
> > +
> > +
> > +	/* Split the block until we reach list "i" */
> > +	for (j--; j > i; j--) {
> Perhaps you could explain your use of the bitmaps a bit better to me ?
> 
> —
> Reply to this email directly or view it on GitHub.
>